### DISCUSSION TITLE
Cannot connect to cassandra db from my application

### DISCUSSION THREAD

**Discussion by vivekab** [community/community#26047] [Actions]
<p>Hi everyone,</p>
<p>I have a go backend application where I try to connect to cassandra db and run some tests. So I need an instance to cassandra db. Referring to this: <a href="https://github.com/marketplace/actions/setup-cassandra-action" rel="nofollow noopener">https://github.com/marketplace/actions/setup-cassandra-action</a> I tried connecting to cassandra using the ip provided or the localhost and none of them seem to work.</p>
<p>An issue <a href="https://github.com/fabasoad/setup-cassandra-action/issues/1" rel="nofollow noopener">https://github.com/fabasoad/setup-cassandra-action/issues/1</a> is already created but seems like there is no response.</p>
<p>Thanks in advance.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/vivekab">@vivekab</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>This is not an official github action, it’s recommended to comment on the issue and <span class="mention">@author</span>, wait for the update.</p>
<p>In the meanwhile, you can try to use the official image and set it as a service(doc <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservices" rel="nofollow noopener">here</a>), code sample as below:</p>
<pre><code class="lang-auto">jobs:
  cassandra:
    services:
      cassandra:
        image: cassandra
        ports:
          - 9042:9042
    runs-on: [ubuntu-latest]
    steps:
      - name: check
        run: |
          docker ps
</code></pre>
<p>Please refer to my workflow for more details.<br>
</p><aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/weide-zhou/ticket13/runs/888672071?check_suite_focus=true#step:5:10" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/52772761/181094970-52858129-0a7a-4359-a103-c55f731260c7.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/weide-zhou/ticket13/runs/888672071?check_suite_focus=true#step:5:10" target="_blank" rel="nofollow noopener">weide-zhou/ticket13</a></h3>

<p>Contribute to weide-zhou/ticket13 development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Thanks.</p>

**Comment by vivekab** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>,</p>
<p>Thanks on the update I tried connecting to the container as per your mentioned guidelines.</p>
<p>But still I have problem connecting to the container though.</p>
<p>when I tried running:</p>
<pre><code class="lang-auto">build:
    name: Build and Test
    services:
      cassandra:
        image: cassandra
        ports:
          - 9042:9042
    runs-on: [self-hosted, linux]
    steps:
    - name: check cassandra installation
      run: |
        docker ps
    - name: create containerId
      run: |
        echo "::set-env name=ContianerID::$(docker ps --format {{.ID}})"
    - name: Create products_keyspace
      run: |
        docker exec $ContianerID cqlsh --debug localhost:9024 --execute="CREATE KEYSPACE products_core WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };"
</code></pre>
<p>It gives me connectivity error</p>
<pre><code class="lang-auto">Traceback (most recent call last):
  File "/opt/cassandra/bin/cqlsh.py", line 2443, in &lt;module&gt;
    main(*read_options(sys.argv[1:], os.environ))
  File "/opt/cassandra/bin/cqlsh.py", line 2421, in main
    encoding=options.encoding)
  File "/opt/cassandra/bin/cqlsh.py", line 485, in __init__
    load_balancing_policy=WhiteListRoundRobinPolicy([self.hostname]),
  File "/opt/cassandra/bin/../lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/policies.py", line 417, in __init__
socket.gaierror: [Errno -2] Name or service not known
</code></pre>
<p>I also tried connecting without giving the host and the port but then its still an error.</p>
<pre><code class="lang-auto">docker exec $ContianerID cqlsh -e "CREATE KEYSPACE products_core WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };"
  shell: /bin/bash -e {0}
  env:
    CASS_DB_KEYSPACE: products_core
    CASS_DB_HOST: cassandra
    CASS_DB_PORT: 9042
    CASS_DB_USER: cassandra
    CASS_DB_PASS: cassandra
    ContianerID: 2af6369a2338
Connection error: ('Unable to connect to any servers', {'127.0.0.1': error(111, "Tried connecting to [('127.0.0.1', 9042)]. Last error: Connection refused")})
##[error]Process completed with exit code 1.
</code></pre>
<p>Am I doing something wrong here? :slightly_frowning_face:</p>

**Comment by dakale** [comment, marked as answer]
<p>This is a common problem thats caused by trying to access the service (cassandra) before it is fully online and ready to handle connections. This is more a matter of how docker works when starting containers in the background, its async. We can add a simple healthcheck so that the rest of the steps downstream dont run until the service is healthy (if the container doesnt have any good tools in it for performing a healthcheck, you could always just <code>sleep 30</code> on the host as a step, for eg)</p>
<p>Heres a quick sample I wrote:</p>
<pre><code class="lang-auto">on:
  push:
    paths:
    - .github/workflows/cassandra.yaml

jobs:
  test:
    runs-on: ubuntu-latest
    services:
      cassandra:
        image: cassandra
        ports:
        - 9042:9042
        options: --health-cmd "cqlsh --debug" --health-interval 5s --health-retries 10
    steps:
    - run: docker ps
    - run: docker exec ${{ job.services.cassandra.id }} cqlsh --debug
</code></pre>
<p>Basically docker will run <code>cqlsh --debug</code> in the container every 5s for up to 10 times to update its “health”, and the runner polls the containers health to block before starting the rest of the job</p>

**Comment by vivekab** [comment]
<p><a class="mention" href="/dakale">@dakale</a> and <a class="mention" href="/weide-zhou">@weide-zhou</a>  That really helps, will be following the same from now on. Thanks.</p>

**Comment by Jasperav** [comment]
<p><a class="mention" href="/vivekab">@vivekab</a>  What .yml file are you using ATM? I copied the answer from <a class="mention" href="/dakale">@dakale</a> but I am unable to create a keyspace or to query anything. This simple .yml file fails to run on github:</p>
<pre><code class="lang-auto">name: CasDB

on: push

env:
  CARGO_TERM_COLOR: always


jobs:
  test:
    runs-on: ubuntu-latest
    services:
      cassandra:
        image: cassandra
        ports:
          - 9042:9042
        options: --health-cmd "cqlsh --debug" --health-interval 5s --health-retries 10
    steps:
      - run: docker ps
      - run: docker exec ${{ job.services.cassandra.id }} cqlsh --debug localhost:9042 --execute="use x;"
</code></pre>
<p>Note: I would expect it to fail with a message like: keyspace <code>x</code> does not exists, but I get this error message:</p>
<blockquote>
<p>Using CQL driver: &lt;module ‘cassandra’ from ‘/opt/cassandra/bin/…/lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/<strong>init</strong>.py’&gt;<br>
Using connect timeout: 5 seconds<br>
Using ‘utf-8’ encoding<br>
Using ssl: False<br>
Traceback (most recent call last):<br>
File “/opt/cassandra/bin/cqlsh.py”, line 2459, in <br>
main(*read_options(sys.argv[1:], os.environ))<br>
File “/opt/cassandra/bin/cqlsh.py”, line 2437, in main<br>
encoding=options.encoding)<br>
File “/opt/cassandra/bin/cqlsh.py”, line 485, in <strong>init</strong><br>
load_balancing_policy=WhiteListRoundRobinPolicy([self.hostname]),<br>
File “/opt/cassandra/bin/…/lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/policies.py”, line 417, in <strong>init</strong><br>
socket.gaierror: [Errno -2] Name or service not known<br>
##[error]Process completed with exit code 1.</p>
</blockquote>

**Comment by Jasperav** [comment]
<p><a class="mention" href="/vivekab">@vivekab</a> could you share your final yml file? I also asked my question on <a href="https://stackoverflow.com/questions/63410396/setup-cassandra-container-in-github-actions-and-query" rel="nofollow noopener">https://stackoverflow.com/questions/63410396/setup-cassandra-container-in-github-actions-and-query</a> and <a href="https://www.reddit.com/r/cassandra/comments/id8lgo/use_cassandra_with_github_actions/" rel="nofollow noopener">https://www.reddit.com/r/cassandra/comments/id8lgo/use_cassandra_with_github_actions/</a> but I don’t get any useful responses</p>

**Comment by vivekab** [comment]
<p>Hi <a class="mention" href="/jasperav">@Jasperav</a>,</p>
<p>Here is the yaml that I use.</p>
<pre><code class="lang-auto">jobs:

  build:
    name: Build and Test
    services:
      cassandra:
        image: cassandra
        ports:
          - 9042:9042
        options: --name cassandra_test --health-cmd "cqlsh --debug" --health-interval 5s --health-retries 10
    runs-on: [self-hosted, linux]
    steps:
    - name: Create products_core keyspace
      run: |
        docker exec cassandra_test cqlsh -u cassandra -p cassandra -e "CREATE KEYSPACE test WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };"

</code></pre>
